presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -0,0 +1,481 @@
+                            table.getTableName(),
+                            partitionName,
+                            path,
+                            blockLocation.getOffset() + chunkOffset,
+                            chunkLength,
+                            schema,
+                            partitionKeys,
+                            addresses,
+                            forceLocalScheduling,
+                            session,
+                            effectivePredicate));
+
+                    chunkOffset += chunkLength;
+                    remainingInitialSplits.decrementAndGet();
+                }
+                checkState(chunkOffset == blockLocation.getLength(), "Error splitting blocks");
+            }
+        }
+        else {
+            // not splittable, use the hosts from the first block if it exists
+            List<HostAddress> addresses = ImmutableList.of();
+            if (blockLocations.length > 0) {
+                addresses = toHostAddress(blockLocations[0].getHosts());
+            }
+
+            builder.add(new HiveSplit(connectorId,
+                    table.getDbName(),
+                    table.getTableName(),
+                    partitionName,
+                    path,
+                    start,
+                    length,
+                    schema,
+                    partitionKeys,
+                    addresses,
+                    forceLocalScheduling,
+                    session,
+                    effectivePredicate));
+        }
+        return builder.build();
+    }
+
+    private boolean getForceLocalScheduling(ConnectorSession session)
+    {
+        String forceLocalScheduling = session.getProperties().get(FORCE_LOCAL_SCHEDULING);
+        if (forceLocalScheduling == null) {
+            return this.forceLocalScheduling;
+        }
+
+        try {
+            return Boolean.valueOf(forceLocalScheduling);
+        }
+        catch (IllegalArgumentException e) {
+            throw new PrestoException(NOT_SUPPORTED, "Invalid Hive session property '" + FORCE_LOCAL_SCHEDULING + "=" + forceLocalScheduling + "'");
+        }
+    }
+
+    private static List<HostAddress> toHostAddress(String[] hosts)
+    {
+        ImmutableList.Builder<HostAddress> builder = ImmutableList.builder();
+        for (String host : hosts) {
+            builder.add(HostAddress.fromString(host));
+        }
+        return builder.build();
+    }
+
+    private static List<HivePartitionKey> getPartitionKeys(Table table, Partition partition)
+    {
+        if (isUnpartitioned(partition)) {
+            return ImmutableList.of();
+        }
+        ImmutableList.Builder<HivePartitionKey> partitionKeys = ImmutableList.builder();
+        List<FieldSchema> keys = table.getPartitionKeys();
+        List<String> values = partition.getValues();
+        checkArgument(keys.size() == values.size(), "Expected %s partition key values, but got %s", keys.size(), values.size());
```

Comment:
Is this an internal invariant? If it could be caused by metastore data, use `PrestoExeption`